migrate entity.js to typescript

lindner opened the pull request:
Addresses #1765

lindner commented:
There's some weird stuff going on here that I don't fully understand.

I added a rawData field at the top level, also make Entity abstract and make key() getter abstract.

I wonder if the data() getter needs to be changed to rawData()?

Should I switch rawData from a field to a getter?

Thanks!

shans commented on runtime/ts/entity.ts:
```diff
@@ -5,15 +5,20 @@
+  // tslint:disable-next-line: no-any
+  protected rawData: any;
+
+  constructor(userIDComponent: string|null) {
```
provide with a default (== null), or change null to void or undefined to capture non-provided userIDComponent case.

lindner replied:
done.

shans commented on runtime/ts/entity.ts:
```diff
@@ -51,11 +59,14 @@ export class Entity {
-  static get type() {
+  abstract get key(): {tag: string, schema: Schema};
+
+  get type() {
```
This needs to remain static.

lindner replied:
static get type() cannot be static because it references `this`.

shans replied:
for static methods, 'this' refers to the class, not the instance.

shans replied:
For static methods, `this` refers to the class rather than the instance.

shans commented on runtime/ts/relation.ts:
```diff
@@ -15,14 +15,20 @@ export class Relation extends Entity {
 
   constructor(...entities) {
-    super();
+    super(undefined);
```
undefined is not null, this will probably stop working when we switch on strict checking. Also, semantically userID is optional, so you shouldn't have to specify anything here.

lindner replied:
fixed

shans commented on runtime/ts/relation.ts:
```diff
@@ -15,14 +15,20 @@ export class Relation extends Entity {
     this.entities = entities;
   }
 
+  // should be rawData???
```
No, you can't have both a field and a getter with the same name. I'm pretty sure this should be `data`.

lindner replied:
Looking at the history of this code it appears that `data` was switched to `rawData` -- I was wondering if this was the case here.

shans commented on runtime/ts/schema.ts:
```diff
@@ -392,10 +390,10 @@ export class Schema {
       }
 
-      static get key() {
+      get key() {
```
this needs to remain a static  method.

shans commented on runtime/ts/entity.ts:
```diff
@@ -51,11 +59,14 @@ export class Entity {
   }
 
-  static get type() {
+  abstract get key(): {tag: string, schema: Schema};
```
This is a static getter in schema generated subclasses. You won't be able to maintain this as an abstract method once you return the static signifier, so there's no point this class being converted to abstract either.

shans commented on runtime/ts/relation.ts:
```diff
@@ -15,14 +15,20 @@ export class Relation extends Entity {
   }
 
+  // TODO should this return something else?
+  get key() {
```
This won't be required once you remove it from the super.

lindner replied:
done

lindner commented:
Might be worth spending a few minutes discussing f2f....  The class layout/use is still not entirely clear to me.

shans commented:
I'll take a closer look and we can discuss next week.

shans commented:
The pattern used here is valid js:

```
class Super {
  static get propertyOfThisClass() {
    return 'super';
  }
  static get propertyContainingPositionalInfo() {
    return this.propertyOfThisClass + ' <- this is the class';
  }
}

class Sub extends Super {
  static get propertyOfThisClass() {
    return 'super';
}

console.log(Super.propertyContainingPositionalInfo); // 'super <- this is the class'
console.log(Sub.propertyContainingPositionalInfo); // 'sub <- this is the class'
```

In our case, subclasses of Entity statically report a type that is derived from the static member key.

shans commented:
The pattern used here is valid both for javaScript and for typeScript:

```
class Super {
  static get propertyOfThisClass() {
    return 'super';
  }
  static get propertyContainingPositionalInfo() {
    return this.propertyOfThisClass + ' <- this is the class';
  }
}

class Sub extends Super {
  static get propertyOfThisClass() {
    return 'super';
}

console.log(Super.propertyContainingPositionalInfo); // 'super <- this is the class'
console.log(Sub.propertyContainingPositionalInfo); // 'sub <- this is the class'
```

In our case, subclasses of Entity statically report a type that is derived from the static member key.

lindner commented on runtime/ts/entity.ts:
```diff
@@ -5,15 +5,20 @@
-  constructor(userIDComponent) {
-    assert(!userIDComponent || userIDComponent.indexOf(':') == -1, 'user IDs must not contain the \':\' character');
+export abstract class Entity {
+  private _userIDComponent?: string;
```
done

lindner commented:
PTAL, this should be better.  Thanks for your patience.

- Reverted back to static usages
- Moved the getter for type() to  schema.ts (`this` usage was only compiling by referencing the same class.)  Also now I know how `this`is used in static getters.  yay
- rawData is defined in the base class, since methods there refer to it.
- Found an type issue with the return type for key(), it was returning schema.toLiteral(), but that was clashing with the type signature for `Type.newEntity`
